Components/Inputs/TLV2544.cs

```diff
@@ -0,0 +1,264 @@
+                case -2: ChSel = Command.SEL_TEST2; break;
+                case -3: ChSel = Command.SEL_TEST3; break;
+                default: ChSel = Command.SEL_CH0; break;
+            }
+            if (this.Config.ConversionMode == ConversionMode.SINGLE_SHOT)
+            {
+                ushort StartRead = DoCommand(ChSel, Long: true);
+                return DoCommand(Command.READ_FIFO);
+            }
+            return 0;
+        }
+
+        /// <summary> Does a 12b read/write with the specified command. </summary>
+        /// <param name="Command"> The command (4 MSb) to send. </param>
+        /// <param name="Data"> The data (12 LSb) to send. </param>
+        /// <param name="Long"> Whether to send out additional SCLK pulses for the ADC to do sampling. </param>
+        /// <returns> The 12b data returned by the device. </returns>
+        private ushort DoCommand(Command Command, ushort Data = 0x000, bool Long = false)
+        {
+            byte[] DataOut;
+            if (!Long) { DataOut = new byte[2]; } // Regular command, no sampling.
+            else
+            {
+                switch (this.Config.ConversionClockSrc) // We are doing sampling, determine the correct number of SCLKs.
+                {
+                    case ConversionClockSrc.SCLK:
+                        DataOut = (this.Config.UseLongSample ? new byte[2 + 4] : new byte[2 + 2]);
+                        break;
+                    case ConversionClockSrc.SCLK_HALF:
+                        DataOut = (this.Config.UseLongSample ? new byte[2 + 7] : new byte[2 + 4]);
+                        break;
+                    case ConversionClockSrc.SCLK_QUARTER:
+                        DataOut = (this.Config.UseLongSample ? new byte[2 + 14] : new byte[2 + 7]);
+                        break;
+                    case ConversionClockSrc.INTERNAL: // We are assuming the time between The two SPI transactions is at least 4us for short, or 8us for long sampling. This seems to always be true on the Pi, but is not guaranteed.
+                        DataOut = new byte[2]; // No need for additional SCLKs.
+                        break;
+                    default:
+                        DataOut = new byte[2];
+                        break;
+                }
+            }
+
+            DataOut[0] = (byte)((((byte)Command << 4) & 0b1111_0000) | ((Data >> 8) & 0b0000_1111));
+            DataOut[1] = (byte)(Data & 0b1111_1111);
+            byte[] DataIn = this.Bus.Write(this.CS, DataOut, DataOut.Length);
+            return (ushort)(Command == Command.READ_CONF ?
+                (((DataIn[0] & 0b0000_1111) << 8) | (DataIn[1])) :
+                (DataIn[0] << 4) | ((DataIn[1] & 0b1111_0000) >> 4));
+        }
+
+        #region Structs and Enums
+        public struct Configuration
+        {
+            public VoltageReference VoltageRef;
+            public bool UseLongSample; // TODO: Check if long sampling is working.
```

Comment:
Have you checked if long sampling is working?

Reply:
I have not, I'll give it a try tonight before release.